### DISCUSSION TITLE
How would one access vector tile data as JSON?

### DISCUSSION THREAD

**Discussion by jayarjo** [maplibre/maplibre-gl-js#665] [Q&A] [answered]
I'd ask it on SO, but I think it is too specific/advanced. Basically what I wonder about is whether there is a way to access a result of vector tile to JSON conversion?

I've inspected `MapEventData` in `sourcedata` and noticed that there's some data stored as ArrayBuffer under `tile.latestRawTileData`, but how would I extract that? Simply convert to text? I believe there should be some kind of utility in the source code already that does that right away. Can you help?

**Comment by HarelM** [comment, marked as answer]
There's a npm package to convert pbf to geojson called vt-geojson.
But I'm not sure I understand the question entirely.
Can you elaborate on the scenario please?

**Reply by jayarjo** [reply]
Thanks. I've also found [vector-tile-js](https://github.com/mapbox/vector-tile-js) - are these interchangeable? I see that that's what mapbox uses itself under the hood.

Regarding the scenario. Basically I need to represent a layer of Points using complex html (proper, fully interactive, but tiny svg chart), rather than static style or an image. So what i currently do is that I style the layer as `symbol` but provide a non-existent image id, in order to intercept it later with `styleimagemissing` and insert a marker instead of that image. However there's only that much data that one can relay via `styleimagemissing`, so I was looking to hook onto `sourcedata` instead and process the layer directly, rather than hacking around.

Can you suggest anything better maybe?

**Reply by HarelM** [reply]
What about `querySourceFeatures`?

**Reply by jayarjo** [reply]
Hm... 🤔

And call it on every `sourcedata`? Do I need to have a layer defined for the source that I want to query?

**Reply by HarelM** [reply]
The following is the angular implementation of something similar to what you are trying to achieve I think:
https://github.com/maplibre/ngx-maplibre-gl/blob/main/projects/ngx-maplibre-gl/src/lib/markers-for-clusters/markers-for-clusters.component.ts
It is an angular component to create html markers on top of the map based on `sourcedata` event.
You can see the clusters in action here - you can see both the clusters and the html markers.
https://israelhiking.osm.org.il/
I don't think there's a need for a layer, but it uses a geojson source and not vector tiles, so it might not be exactly what you need...
